File: Android/app/src/main/java/com/dreamproject/inwords/presentation/viewScenario/main/MainFragment.java
package com.dreamproject.inwords.presentation.viewScenario.main;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import com.dreamproject.inwords.R;
import com.dreamproject.inwords.presentation.viewScenario.FragmentWithViewModelAndNav;
import com.jakewharton.rxbinding2.view.RxView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;


/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link MainFragment.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link MainFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class MainFragment extends FragmentWithViewModelAndNav<MainViewModel, MainViewModelFactory> {
    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_PARAM1 = "param1";

    // TODO: Rename and change types of parameters
    private String mParam1;

    private OnFragmentInteractionListener mListener;

    public MainFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @return A new instance of fragment MainFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static MainFragment newInstance(String param1, String param2) {
        MainFragment fragment = new MainFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null)
            mParam1 = getArguments().getString(ARG_PARAM1);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        Button buttonGoLogin = view.findViewById(R.id.buttonGoLogin);
        Button buttonGetWordList = view.findViewById(R.id.buttonGetWordList);

        buttonGoLogin.setOnClickListener(
                Navigation.createNavigateOnClickListener(R.id.action_mainFragment_to_loginFragment, null));
//TODO это не по паттерну. перенести навигацию в модель. но это пока костыль так что оставить как есть
        viewModel.onGetAllHandler(RxView.clicks(buttonGetWordList));

        //BottomNavigationView navigation = getActivity().findViewById(R.id.navigation);
        //viewModel.navigationItemSelectionHandler(RxBottomNavigationView.itemSelections(navigation));
    }

    @Override
    protected int getLayout() {
        return R.layout.fragment_main;
    }

    @NonNull
    @Override
    protected Class<MainViewModel> getClassType() {
        return MainViewModel.class;
    }

    // TODO: Rename method, update argument and hook method into UI event
    public void onButtonPressed(Uri uri) {
        if (mListener != null) {
            mListener.onFragmentInteraction(uri);
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        // TODO: Update argument type and name
        void onFragmentInteraction(Uri uri);
    }
}

File: Android/app/src/main/java/com/dreamproject/inwords/presentation/viewScenario/main/MainFragment.kt
package com.dreamproject.inwords.presentation.viewScenario.main

import android.net.Uri
import android.os.Bundle
import android.view.View
import androidx.navigation.Navigation
import com.dreamproject.inwords.R
import com.dreamproject.inwords.core.util.SchedulersFacade
import com.dreamproject.inwords.data.dto.User
import com.dreamproject.inwords.data.dto.noUser
import com.dreamproject.inwords.presentation.viewScenario.FragmentWithViewModelAndNav
import com.facebook.imagepipeline.request.ImageRequestBuilder
import com.jakewharton.rxbinding2.view.RxView
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.card_dictionary.view.*
import kotlinx.android.synthetic.main.card_profile.view.*
import kotlinx.android.synthetic.main.fragment_home.*

class MainFragment : FragmentWithViewModelAndNav<MainViewModel, MainViewModelFactory>() {
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

//TODO это не дело. перенести навигацию в модель. но это пока костыль
        viewModel.onGetAllHandler(RxView.clicks(buttonGetWordList))

        compositeDisposable.add(subscribeProfile())
        compositeDisposable.add(subscribeDictionary())
        //BottomNavigationView navigation = getActivity().findViewById(R.id.navigation);
        //viewModel.navigationItemSelectionHandler(RxBottomNavigationView.itemSelections(navigation));
    }

    private fun toLoginClickListener() = Navigation
            .createNavigateOnClickListener(R.id.action_mainFragment_to_loginFragment, null)

    private fun subscribeProfile(): Disposable {
        fun showProfile() {
            profile.avatar.visibility = View.VISIBLE
            profile.experience.visibility = View.VISIBLE
            profile.name.visibility = View.VISIBLE
            profile.clickToLogin.visibility = View.GONE

            profile.setOnClickListener(toLoginClickListener()) //TODO show profile listener here
        }

        fun showLoginHint() {
            profile.avatar.visibility = View.GONE
            profile.experience.visibility = View.GONE
            profile.name.visibility = View.GONE
            profile.clickToLogin.visibility = View.VISIBLE

            profile.setOnClickListener(toLoginClickListener())
        }

        fun renderUser(user: User) {
            if (user.avatar != null) {
                val request = ImageRequestBuilder.newBuilderWithSource(Uri.parse(user.avatar))
                        .setProgressiveRenderingEnabled(true)
                        .setLocalThumbnailPreviewsEnabled(true)
                        .build()

                profile.avatar.setImageRequest(request)
            }

            profile.experience.text = getString(R.string.user_experience, 15)
            profile.name.text = user.userName

            showProfile()
        }

        fun renderNoUser() {
            val hintText = getString(R.string.sign_up_proposal)
            profile.name.text = hintText
            profile.experience.text = "Войдите в аккаунт, чтобы использовать приложение по полной"

            showLoginHint()
        }

        fun renderSuccess(user: User) {
            if (user.userId == noUser.userId) {
                renderNoUser()
            } else {
                renderUser(user)
            }

            profile.visibility = View.VISIBLE
        }

        fun renderLoading() {
            val loadingText = getString(R.string.loading_text_placeholder)
            profile.name.text = loadingText
            profile.experience.text = loadingText

            showProfile()
        }

        return viewModel.profileDataSubject
                .observeOn(SchedulersFacade.ui())
                .doOnSubscribe { renderLoading() }
                .subscribe(::renderSuccess)
    }

    private fun subscribeDictionary(): Disposable {
        fun applyListener() {
            dictionary.setOnClickListener { navController.navigate(R.id.action_mainFragment_to_translationMainFragment) }
        }

        fun renderCount(count: Int) {
            dictionary.dictSize.text = getString(R.string.words_in_dictionary, count)
        }

        fun renderError() {
            val errorText = getString(R.string.error_text_placeholder)
            dictionary.dictSize.text = errorText
        }

        return viewModel.wordsCountSubject
                .observeOn(SchedulersFacade.ui())
                .doOnSubscribe { applyListener() }
                .subscribe(::renderCount) { renderError() }
    }

    override fun getLayout(): Int {
        return R.layout.fragment_home
    }

    override fun getClassType(): Class<MainViewModel> {
        return MainViewModel::class.java
    }
}
